A77-xds-rate-limiting-rlqs.md

```diff
@@ -80,113 +83,383 @@ config:
+
+The RLQS Bucket tracks the current rate limiting assignment and the quota
+usage for a specific bucket, identified by a `bucket_id`. It uses this data to
+determine whether to allow or deny a request. The bucket also provides a
+thread-safe mechanism to snapshot and reset its quota usage when building usage
+reports.
+
+RLQS Bucket object will include the following data members:
+
+-   `bucket_id`: From filter config and RPC metadata, constant, initialized at
+    bucket creation. A unique identifier for the bucket.
+-   "No Assignment" and "Expired Assignment" strategies: Initialized at first
+    RPC for the bucket from filter config, constant. Fallback rate limiting
+    strategies.
+-   Active Assignment: Initialized at first RPC for the bucket from filter
+    config, updated on RLQS server responses. The current rate limiting strategy
+    to apply to requests, and associated expiration timestamps.
+-   Request Counters: Initialized at 0, updated on data plane RPCs, reset on
+    report timers and RLQS server responses. Tracks the number allowed/denied
+    requests for the bucket.
+
+##### RLQS Buckets and Multithreading
+
+There are several mutex-synchronized operations on RLQS buckets that are
+executed during latency-sensitive data plane RPC processing:
+
+1. Inserting/reading a bucket from the RLQS Bucket Map using `bucket_id`. Note
+   that `bucket_id` is represented as a `Map<String, String>`, which may
+   introduce complexities in efficient cache sharding for certain programming
+   languages.
+2. Incrementing `num_request_allowed`/`num_request_denied` bucket counters.
+
+Each gRPC implementation needs to consider what synchronization primitives are
+available in their language to minimize the thread lock time.
+
+#### RLQS Client
+
+The RLQS Client is responsible for communication with the Rate Limit Quota
+Service (RLQS) server. It manages a gRPC stream to the RLQS server, used for
+sending periodic bucket usage reports and receiving new rate limit quota
+assignments. The client handles reconnects and manages the lifecycle of the
+stream.
+
+RLQS Client object will include the following data members:
+
+-   gRPC Channel: From filter config, initialized at instantiation, constant. A
+    channel to the RLQS server.
+-   gRPC Stream: Initialized at instantiation, replaced any time the stream
+    fails (with backoff). Accessed when we get the first data plane RPC for a
+    given bucket and when a report timer fires. A bidirectional gRPC stream to
+    the RLQS server.
+-   Buckets Update Callback: Provided by RLQS Filter state at instantiation,
+    constant, accessed on responses from the RLQS server. A callback to notify
+    the RLQS Filter State of updates to quota assignments.
+
+### Handling Events
+
+#### On LDS/RDS Updates
+
+When receiving an LDS/RDS update, the RLQS filter will perform the following 
+steps for each route:
+
+1.  Parse the new filter config and per-route overrides into an internal RLQS
```

Comment:
One thing that seems to be missing from this gRFC is config validation information.  We need to list all of the fields in the RLQS filter config that we support, and we need to say what validation we are going to do on each field.  In particular, we need to spell out which values will cause us to NACK.

Reply:
One particular thing of note in the context of config validation: When validating unified matcher trees, we will need to impose some max recursion depth limit, to avoid OOMing due to a control plane sending us too deep of a tree.

Reply:
This is done.

Reply:
Wrong link. 

- The filter config and validation is in https://github.com/sergiitk/grfc-proposal/blob/a77-rlqs/A77-xds-rate-limiting-rlqs.md#filter-configuration
- Unified matcher parsing and validation, including the tree depth restrictions are in https://github.com/sergiitk/grfc-proposal/blob/a77-rlqs/A77-xds-rate-limiting-rlqs.md#unified-matcher-matcher